gateway/dataimpl/server_v1/kvserver.go

```diff
@@ -1394,156 +1394,43 @@ func (s *KvServer) GetAllReplicas(in *kv_v1.GetAllReplicasRequest, out kv_v1.KvS
-			}
-
-			s.logger.Debug("GetAllReplicas GetFromReplica failed but is being ignored",
-				zap.Uint32("replicaIdx", replicaIdx),
-				zap.Error(err))
-			return nil, nil
-		}
-
-		return &kv_v1.GetAllReplicasResponse{
-			IsReplica:    true,
-			Content:      result.Value,
-			ContentFlags: result.Flags,
-			Cas:          result.Cas,
-		}, nil
-	}
-
-	// our current implementation of GetAllReplicas is somewhat less optimal compared
-	// to what should be possible with full gocbcorex support for this.  primarily, we
-	// execute a replica read on all _possible_ replicas, even if we don't have that
-	// many replicas configured.  Any replicas which are not accessible will return
-	// an ErrInvalidReplicaIdx error, and since all errors are ignored by GetReplicas,
-	// we will simply ignore sending that particular value.
-
-	maxReplicaScans := uint32(3)
-
-	type result struct {
-		res   *kv_v1.GetAllReplicasResponse
-		errSt *status.Status
-	}
-	outCh := make(chan result, maxReplicaScans+1)
-
-	go func() {
-		res, errSt := getFromMaster()
-		outCh <- result{
-			res:   res,
-			errSt: errSt,
-		}
-	}()
-	for replicaIdx := uint32(0); replicaIdx < maxReplicaScans; replicaIdx++ {
-		go func(replicaIdx uint32) {
-			res, errSt := getFromReplica(replicaIdx)
-			outCh <- result{
-				res:   res,
-				errSt: errSt,
-			}
-		}(replicaIdx)
-	}
-
-	remainingReads := 1 + maxReplicaScans
-	asyncReadRemaining := func() {
-		go func() {
-			for remainingReads > 0 {
-				<-outCh
-				remainingReads--
-			}
-			close(outCh)
-		}()
-	}
-
-	// Errors for the get requests are ignored unless they are in a specific set of errors.
-	// This means that we need to bail out when remainingReads is 0 - it's possible for all results
-	// to not write an error or a result into the result written to the channel.
-	for remainingReads > 0 {
-		firstRes := <-outCh
-		remainingReads--
-
-		if firstRes.errSt != nil {
-			// if the first result had some sort of error, we start a goroutine to clean
-			// up the remaining results, and immediately return the generated error.
-
-			asyncReadRemaining()
-			return firstRes.errSt.Err()
+			return err
 		}
 
-		if firstRes.res != nil {
-			// If we actually got a proper result from this response, we send it and then
-			// break to the remaining handling below, otherwise we fetch the next result.
-			err := out.Send(firstRes.res)
-			if err != nil {
-				asyncReadRemaining()
-				return s.errorHandler.NewGenericStatus(err).Err()
-			}
-
-			break
-		}
+		res, err = replicaStream.Next()
 	}
 
-	// once the first read has been successful, we no longer accept errors and simply
-	// pretend like they did not occur
-	for remainingReads > 0 {
-		nextRes := <-outCh
-		remainingReads--
-
-		if nextRes.res != nil {
-			err := out.Send(nextRes.res)
-			if err != nil {
-				asyncReadRemaining()
-				return s.errorHandler.NewGenericStatus(err).Err()
-			}
-		}
+	if err != nil {
```

Comment:
I think this is a change in error handling behaviour? Previously if the first read failed then we'd return an error. I also think we should just put the error checking blocks next to the error assignment. It's a bit more verbose but much easier to read/understand intent.

Reply:
Doesn't this still return an error on the first read failure?  Since the for loop above is dependent on `res=nil`?

Reply:
Ah I misunderstood this. I do find it a bit hard to follow and think we should just use a for {} loop without calling .Next() before it. Then we can just handle the errors next to the error being returned.

Reply:
Filed https://couchbasecloud.atlassian.net/browse/ING-804 for future improvements to this.